### PR TITLE
Validate module names during creation (#609)

### PR DESCRIPTION
**Pull request by pq**
Same UX as described in closed PR (https://github.com/flutter/flutter-intellij/pull/724) but without the build dependency issues. 👍 

Fixes: #609 

@devoncarew @skybrian

### CODE REVIEW

File: src/io/flutter/FlutterUtils.java [hunk 1]
```diff
@@ -58,4 +61,15 @@ public static VirtualFile getRealVirtualFile(@Nullable PsiFile psiFile) {
...
   }
+
+  /**
+   * Checks whether a given string is a valid Dart identifier.
+   *
+   * See: https://www.dartlang.org/guides/language/spec
+   * @param id the string to check
+   * @return true if a valid identifer, false otherwise.
+   */
+  public static boolean isValidDartdentifier(@NotNull String id) {
```
Comment (skybrian) [inline src/io/flutter/FlutterUtils.java]: Do you want to check for Dart keywords?
Here is the list we use in the protobuf code generator:
https://github.com/dart-lang/dart-protoc-plugin/blob/master/lib/names.dart#L246

Or if we don't care whether it's a Dart keyword, perhaps this method should be renamed?

File: src/io/flutter/module/FlutterModuleBuilder.java [hunk 2]
```diff
@@ -107,6 +109,21 @@ private static void createProjectFiles(@NotNull final ModifiableRootModel model,
     }
   }
 
+  @Override
+  public boolean validateModuleName(@NotNull String moduleName) throws ConfigurationException {
+    if (!FlutterUtils.isValidDartdentifier(moduleName)) {
+      throw new ConfigurationException("Invalid Flutter module name '" + moduleName + "'; must be a valid Dart identifier.");
```
Comment (skybrian) [inline src/io/flutter/module/FlutterModuleBuilder.java]: It's not obvious why the IDEA module name needs to be a valid Dart identifier. Add a comment explaining this?

File: src/io/flutter/FlutterUtils.java (hunk 1 above)
Comment (devoncarew) [inline src/io/flutter/FlutterUtils.java]: In the past we had issues w/ users choosing project names that were the same as common pub package names (esp. the names of flutter dependencies), and running into issues w/ the first `pub get`. For `flutter create`, we proscribe the common package names: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/create.dart#L248.

Comment (devoncarew) [inline src/io/flutter/FlutterUtils.java]: (`isValidDartdentifier` ==> `isValidDartIdentifier`)

**Comment by pq** [comment]
Lots of good comments.  Thanks!  I'll follow-up with a PR in a bit. 👍

**Comment by pq** [comment]
OK, some fresh updates.  Thanks again for all the feedback.

PTAL.

@devoncarew @skybrian

File: src/io/flutter/module/FlutterModuleBuilder.java [hunk 3]
```diff
@@ -160,6 +162,32 @@ public boolean validate(Project current, Project dest) {
...
+    if (FlutterUtils.isDartKeword(moduleName)) {
+      throw new ConfigurationException("Invalid module name: '" + moduleName + "' - must not be a Dart keyword.");
+    }
+
+    if (!FlutterUtils.isValidDartIdentifier(moduleName)) {
+      throw new ConfigurationException("Invalid module name: '" + moduleName + "' - must be a valid Dart identifier.");
+    }
+
+    if (FlutterConstants.FLUTTER_PACKAGE_DEPENDENCIES.contains(moduleName)) {
+      throw new ConfigurationException("Invalid module name: '" + moduleName + "' - this will conflict with Flutter package dependencies.");
```
Comment (devoncarew) [inline src/io/flutter/module/FlutterModuleBuilder.java]: 👍 nice messages :)

**Comment by devoncarew** [comment]
lgtm

**Pull request merged**